### PR TITLE
mdbook setup

### LINKED ISSUE #64: put link to html version of book in README
I know I'm an idiot who can't use the internet, but search for me stubbornly refuses to give me the URL of the web-based version of this book. I'm always being linked to the github. Can you please put a link in the README of the repo to the web URL? I would do it myself but I can't find it :-( Even just a link like they do with mathlib (in "About in top right of the github main page) would be great!

### PR DESCRIPTION
**Pull request by Seasawher**
resolve #64

I have used mdbook to host this book as web pages.

For deployment, please use github actions instead of branch deploying.

### CODE REVIEW

**Comment by Julian** [comment]
This idea in theory seems reasonable to me but I think there's a few things to work out.

Specifically if we do this, it seems like we should avoid duplicating the README in the SUMMARY.md, as they'll easily go out of sync.

Both *would* be being used for similar purposes -- the README, besides being human-readable, is also used [for indicating the order of pages for the PDF](https://github.com/leanprover-community/lean4-metaprogramming-book/blob/master/.github/workflows/book.yml#L20-L21) which we build via `pandoc`. It'd be nice to have one spot we use for that rather than relying on needing to update both places with the same information.

Moreover, I think mdbook still doesn't support outputting PDF files (https://github.com/rust-lang/mdBook/issues/815), and really if we do use something like `mdbook` it'd be nice if it were something that supports both HTML and PDF without additional tricks. In that issue thread are a few other linked tools -- do you have any experience with them? Specifically `mdbook-typst` looks nice at first glance. Perhaps it's worth trying to see if it can handle the PDF build?

(P.S. -- If we do get to merging this, we'll need help from someone with full permissions on the repository, as I don't see the Settings tab here. We can ping them if/when we're ready to merge I think.

**Comment by Seasawher** [comment]
> Specifically mdbook-typst looks nice at first glance. Perhaps it's worth trying to see if it can handle the PDF build?

I tried `mdbook-typst`. Sure, it could build a PDF. But I think `mdbook-typst` is not suitable for our needs... I would prefer to use the existing workflow of building and releasing PDFs.

the reasons are:
* The style of the PDF changes and makes it look unpolished.
* Duplication of README and SUMMARY content is not a significant problem.
* mdbook-typst seemed unstable and did not work as documented.

**Comment by Julian** [comment]
OK -- I'm fine with it if it didn't work out (thanks for trying), though I don't really agree with duplication not being a significant problem -- I think no one will remember/know to update both when contributing changes as-is. At very minimum we need to clearly write in the README that updates need to go in both places, but even better would likely be symlinking one to the other.

**Comment by Seasawher** [comment]
I think mdbook-pandoc has more potential. However, I don't think we should change the PDF output method in this pull request.

**Comment by Seasawher** [comment]
What about using a separate workflow to generate one from the other, Readme and Summary?
Or how about using a workflow to verify that these match?
I think the above suggestion would be easier than building the pdf in mdbook.

**Comment by Julian** [comment]
All of the above are also OK with me (including not doing them in this specific PR!)

**Comment by Seasawher** [comment]
Thank you very much. In this pull request, I will concentrate on adding the ability to host web pages in mdbook.
I would like to work on the match check between readme and summary as a separate issue.

**Comment by Julian** [comment]
Cool, thanks again for the PR and the work! I think this is basically ready to merge then after you back out the `typst` additions to the workflow and files!

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      - name: checkout
+        uses: actions/checkout@v4
+
+      - name: install elan
+        run: |
+          curl https://raw.githubusercontent.com/leanprover/elan/master/elan-init.sh -sSf | sh -s -- -y --default-toolchain $(cat lean-toolchain)
+          echo "$HOME/.elan/bin" >> $GITHUB_PATH
+
+      - name: build markdown files by mdgen
+        run: lake exe mdgen lean md
```
Comment (Julian) [inline .github/workflows/deploy.yml:34]:
```suggestion
        run: lake run build
```

Comment (Seasawher) [reply .github/workflows/deploy.yml:34]: `lake run build` raises an error in GitHub Action (error message said that external command was not allowed..). so this should be as it is

Comment (Julian) [reply .github/workflows/deploy.yml:34]: Interesting! Does that not happen locally when running it? I'd check myself but I have some issues building projects at the moment it seems (which are related to my setup not mdgen)

Comment (Seasawher) [reply .github/workflows/deploy.yml:34]: There were no errors when I ran `lake run build` locally.

**Comment by Seasawher** [comment]
@Julian I have set up the workflow. Please check it out.

**Comment by Seasawher** [comment]
This PR also resolves #113.

**Comment by Julian** [comment]
Great! This looks mergeable to me. Well done. Now we need someone to enable pages I think, assuming it's ok with folks that this then lives on https://leanprover-community.github.io. @jcommelin / @PatrickMassot / @eric-wieser  / @semorrison assuming one of you have higher perms on this repository and are OK with doing so (making the book available at https://leanprover-community.github.io/lean4-metaprogramming-book/) can you tick the box in Settings to enable GitHub pages?

**Comment by PatrickMassot** [comment]
I did it, guessing you want to deploy using GitHub actions.

**Comment by Julian** [comment]
Great, thanks Patrick!

@Seasawher let's see how it goes, hitting the big green button.

**Comment by Julian** [comment]
Nice! Looks like it built! https://leanprover-community.github.io/lean4-metaprogramming-book/ is live. Thanks again!

**Comment by Seasawher** [comment]
@Julian 

Thank you very much!

It would be better if the URL was written in a place where it could be easily found, e.g. in the description section of the repository.

**Pull request merged**